File: web/packages/core/src/blob.ts
/**
 * Blob frames — the one binary framing of every `blob` bulk channel: a named, sized blob sent as
 * offset-ordered chunks that an envelope refers to. The chunk codec, the reference guard and the
 * receive-side store (either arrival order, bounded, timed out) live here so no capability
 * re-implements them.
 * spec: docs/08-protocol.md#blob-frames
 */
import { FjarrError } from "./errors.js";

export const BLOB_VERSION = 1;
export const BLOB_HEADER_BYTES = 37; // version(1) id(16) offset(8) len(8) payload_len(4)
export const BLOB_MAX_CHUNK = 256 * 1024;
/** Whole-blob reassembly cap for `receive()`. */
export const BLOB_MAX_BYTES = 16 * 1024 * 1024;
/** Incomplete or unclaimed blobs kept per channel (oldest evicted). */
export const BLOB_PENDING_BYTES = 8 * 1024 * 1024;
/** A reference whose blob has not completed by then fails; an unclaimed blob is forgotten. */
export const BLOB_PENDING_TTL_MS = 30_000;

/** `{"blob": id, "len": n, "type": media type}` — the `blob-ref` fragment of the envelope schema. */
export interface BlobRef {
  blob: string;
  len: number;
  type?: string;
}

const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function isBlobRef(v: unknown): v is BlobRef {
  if (!v || typeof v !== "object") return false;
  const r = v as Record<string, unknown>;
  return typeof r.blob === "string" && UUID.test(r.blob) && typeof r.len === "number" && Number.isInteger(r.len) && r.len >= 0 && (r.type === undefined || typeof r.type === "string");
}

export interface BlobChunk {
  blobId: string;
  offset: number;
  blobLen: number;
  payload: Uint8Array;
}

function uuidBytes(text: string): Uint8Array | null {
  if (!UUID.test(text)) return null;
  const hex = text.replace(/-/g, "");
  const out = new Uint8Array(16);
  for (let i = 0; i < 16; i++) out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  return out;
}

function uuidText(bytes: Uint8Array): string {
  let hex = "";
  for (let i = 0; i < 16; i++) hex += bytes[i]!.toString(16).padStart(2, "0");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

/** One frame: header + payload. Throws on a malformed id or a chunk that overruns the blob. */
export function encodeBlobChunk(blobId: string, offset: number, blobLen: number, payload: Uint8Array): Uint8Array {
  const id = uuidBytes(blobId);
  if (!id) throw new FjarrError("payload-invalid", `blob id is not a uuid: ${blobId}`);
  if (payload.byteLength > BLOB_MAX_CHUNK) throw new FjarrError("payload-invalid", "blob chunk above 256 KiB (docs/08)");
  if (offset + payload.byteLength > blobLen) throw new FjarrError("payload-invalid", "blob chunk overruns blob_len");
  const frame = new Uint8Array(BLOB_HEADER_BYTES + payload.byteLength);
  const view = new DataView(frame.buffer);
  frame[0] = BLOB_VERSION;
  frame.set(id, 1);
  view.setBigUint64(17, BigInt(offset));
  view.setBigUint64(25, BigInt(blobLen));
  view.setUint32(33, payload.byteLength);
  frame.set(payload, BLOB_HEADER_BYTES);
  return frame;
}

/** Parse a frame; null when the version, the lengths or the offset disagree (dropped and counted by the caller). */
export function parseBlobChunk(frame: ArrayBuffer | Uint8Array): BlobChunk | null {
  const bytes = frame instanceof Uint8Array ? frame : new Uint8Array(frame);
  if (bytes.byteLength < BLOB_HEADER_BYTES || bytes[0] !== BLOB_VERSION) return null;
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const offset = Number(view.getBigUint64(17));
  const blobLen = Number(view.getBigUint64(25));
  const payloadLen = view.getUint32(33);
  if (payloadLen !== bytes.byteLength - BLOB_HEADER_BYTES) return null;
  if (payloadLen > BLOB_MAX_CHUNK) return null;
  if (offset > blobLen || payloadLen > blobLen - offset) return null;
  return { blobId: uuidText(bytes.subarray(1, 17)), offset, blobLen, payload: bytes.subarray(BLOB_HEADER_BYTES) };
}

/** Split bytes into chunks of at most `chunkBytes` payload (an SCTP message minus the header). */
export function* blobChunks(blobId: string, bytes: Uint8Array, chunkBytes: number): Generator<Uint8Array> {
  const n = Math.min(chunkBytes, BLOB_MAX_CHUNK);
  if (bytes.byteLength === 0) {
    yield encodeBlobChunk(blobId, 0, 0, bytes);
    return;
  }
  for (let off = 0; off < bytes.byteLength; off += n) yield encodeBlobChunk(blobId, off, bytes.byteLength, bytes.subarray(off, Math.min(off + n, bytes.byteLength)));
}

interface Entry {
  parts: Uint8Array[];
  received: number;
  len: number;
  complete: boolean;
  first: number;
  last: number;
}

interface Waiter {
  ref: BlobRef;
  resolve: (bytes: Uint8Array) => void;
  reject: (e: Error) => void;
  timer: ReturnType<typeof setTimeout>;
}

export interface BlobReceiverOptions {
  maxBlob?: number;
  pendingBytes?: number;
  ttlMs?: number;
  now?: () => number;
}

/**
 * The receive side of one capability's blob channel: collects chunks (either arrival order relative
 * to the referencing envelope), keeps unclaimed blobs under the docs/08 bounds and resolves
 * `receive(ref)` once the whole blob is there — or rejects after the timeout, on a mismatching
 * length, or when the channel closes.
 */
export class BlobReceiver {
  private readonly entries = new Map<string, Entry>();
  private readonly waiters = new Map<string, Waiter>();
  private readonly chunkHandlers = new Set<(chunk: BlobChunk) => void>();
  private bytes = 0;
  private droppedCount = 0;
  private readonly maxBlob: number;
  private readonly limit: number;
  private readonly ttl: number;
  private readonly now: () => number;

  constructor(options: BlobReceiverOptions = {}) {
    this.maxBlob = options.maxBlob ?? BLOB_MAX_BYTES;
    this.limit = options.pendingBytes ?? BLOB_PENDING_BYTES;
    this.ttl = options.ttlMs ?? BLOB_PENDING_TTL_MS;
    this.now = options.now ?? (() => Date.now());
  }

  get dropped(): number {
    return this.droppedCount;
  }
  get pendingBytes(): number {
    return this.bytes;
  }
  get size(): number {
    return this.entries.size;
  }

  /** Streaming form for large blobs (files, M4): every validated chunk, in order. */
  onChunk(handler: (chunk: BlobChunk) => void): () => void {
    this.chunkHandlers.add(handler);
    return () => {
      this.chunkHandlers.delete(handler);
    };
  }

  /** A binary message from the channel. Returns the id of the blob it completed, if any. */
  onMessage(data: ArrayBuffer | Uint8Array): string | null {
    const chunk = parseBlobChunk(data);
    if (!chunk) {
      this.droppedCount++;
      return null;
    }
    for (const h of Array.from(this.chunkHandlers)) h(chunk);
    return this.collect(chunk);
  }

  private collect(chunk: BlobChunk): string | null {
    const now = this.now();
    this.expire(now);
    if (chunk.blobLen > this.maxBlob) {
      this.droppedCount++;
      return null;
    }
    let e = this.entries.get(chunk.blobId);
    if (!e) {
      if (chunk.offset !== 0) {
        this.droppedCount++;
        return null;
      }
      e = { parts: [], received: 0, len: chunk.blobLen, complete: false, first: now, last: now };
      this.entries.set(chunk.blobId, e);
    }
    if (e.complete || chunk.blobLen !== e.len || chunk.offset !== e.received) {
      this.droppedCount++;
      this.discard(chunk.blobId);
      return null;
    }
    e.parts.push(chunk.payload.slice());
    e.received += chunk.payload.byteLength;
    e.last = now;
    this.bytes += chunk.payload.byteLength;
    e.complete = e.received === e.len;
    while (this.bytes > this.limit && this.entries.size > 1) {
      let oldest: string | null = null;
      for (const [id, other] of this.entries) if (id !== chunk.blobId && (oldest === null || other.first < this.entries.get(oldest)!.first)) oldest = id;
      if (oldest === null) break;
      this.droppedCount++;
      this.discard(oldest);
    }
    if (!e.complete) return null;
    const waiter = this.waiters.get(chunk.blobId);
    if (waiter) this.settle(chunk.blobId, waiter);
    return chunk.blobId;
  }

  private settle(id: string, waiter: Waiter): void {
    const bytes = this.take(id);
    this.waiters.delete(id);
    clearTimeout(waiter.timer);
    if (!bytes) return;
    if (bytes.byteLength !== waiter.ref.len) {
      waiter.reject(new FjarrError("payload-invalid", `blob ${id}: ${bytes.byteLength} bytes arrived, the reference said ${waiter.ref.len}`));
      return;
    }
    waiter.resolve(bytes);
  }

  /** The bytes of a complete blob, removed from the store; null if unknown or incomplete. */
  take(id: string): Uint8Array | null {
    const e = this.entries.get(id);
    if (!e || !e.complete) return null;
    this.discard(id);
    if (e.parts.length === 1) return e.parts[0]!;
    const out = new Uint8Array(e.len);
    let off = 0;
    for (const p of e.parts) {
      out.set(p, off);
      off += p.byteLength;
    }
    return out;
  }

  discard(id: string): void {
    const e = this.entries.get(id);
    if (!e) return;
    this.bytes -= e.received;
    this.entries.delete(id);
  }

  expire(now = this.now()): void {
    for (const [id, e] of Array.from(this.entries)) {
      if (now - e.last > this.ttl && !this.waiters.has(id)) {
        this.droppedCount++;
        this.discard(id);
      }
    }
  }

  /** Resolve a reference found in an envelope: now if the blob already arrived, else when it completes. */
  receive(ref: BlobRef, options: { signal?: AbortSignal; timeoutMs?: number } = {}): Promise<Uint8Array> {
    if (!isBlobRef(ref)) return Promise.reject(new FjarrError("payload-invalid", "not a blob reference"));
    if (ref.len > this.maxBlob) return Promise.reject(new FjarrError("payload-invalid", `blob ${ref.blob}: ${ref.len} bytes is above the ${this.maxBlob}-byte reassembly cap`));
    const existing = this.waiters.get(ref.blob);
    if (existing) return Promise.reject(new FjarrError("payload-invalid", `blob ${ref.blob} is already being received`));
    return new Promise<Uint8Array>((resolve, reject) => {
      const e = this.entries.get(ref.blob);
      if (e?.complete) {
        const bytes = this.take(ref.blob)!;
        if (bytes.byteLength !== ref.len) reject(new FjarrError("payload-invalid", `blob ${ref.blob}: ${bytes.byteLength} bytes arrived, the reference said ${ref.len}`));
        else resolve(bytes);
        return;
      }
      const timer = setTimeout(() => {
        this.waiters.delete(ref.blob);
        this.discard(ref.blob);
        reject(new FjarrError("timeout", `blob ${ref.blob} did not complete within ${options.timeoutMs ?? this.ttl} ms`));
      }, options.timeoutMs ?? this.ttl);
      const waiter: Waiter = { ref, resolve, reject, timer };
      this.waiters.set(ref.blob, waiter);
      options.signal?.addEventListener(
        "abort",
        () => {
          if (this.waiters.get(ref.blob) !== waiter) return;
          this.waiters.delete(ref.blob);
          clearTimeout(timer);
          reject(new FjarrError("closed", "aborted"));
        },
        { once: true },
      );
    });
  }

  /** The channel closed or the peer is gone: every waiter gives up, nothing is kept. */
  close(reason = "bulk channel closed"): void {
    for (const [id, w] of Array.from(this.waiters)) {
      clearTimeout(w.timer);
      this.waiters.delete(id);
      w.reject(new FjarrError("closed", `blob ${id}: ${reason}`));
    }
    this.entries.clear();
    this.bytes = 0;
  }
}

File: agent/src/core/blob.cpp
// spec: docs/08-protocol.md#blob-frames
#include <fjarr/blob.hpp>

#include <algorithm>
#include <cstring>

#include <fjarr/errors.hpp>

namespace fjarr::blob {

namespace {
int hex_value(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    if (c >= 'a' && c <= 'f') return c - 'a' + 10;
    if (c >= 'A' && c <= 'F') return c - 'A' + 10;
    return -1;
}
void put_u64(std::byte *out, std::uint64_t v) {
    for (int i = 7; i >= 0; i--) {
        out[i] = static_cast<std::byte>(v & 0xff);
        v >>= 8;
    }
}
void put_u32(std::byte *out, std::uint32_t v) {
    for (int i = 3; i >= 0; i--) {
        out[i] = static_cast<std::byte>(v & 0xff);
        v >>= 8;
    }
}
std::uint64_t get_u64(const std::byte *in) {
    std::uint64_t v = 0;
    for (int i = 0; i < 8; i++) v = (v << 8) | static_cast<std::uint8_t>(in[i]);
    return v;
}
std::uint32_t get_u32(const std::byte *in) {
    std::uint32_t v = 0;
    for (int i = 0; i < 4; i++) v = (v << 8) | static_cast<std::uint8_t>(in[i]);
    return v;
}
} // namespace

nlohmann::json BlobRef::to_json() const { return nlohmann::json{{"blob", id}, {"len", len}, {"type", type}}; }

std::optional<BlobRef> BlobRef::from_json(const nlohmann::json &j) {
    if (!j.is_object() || !j.contains("blob") || !j["blob"].is_string() || !j.contains("len") || !j["len"].is_number_unsigned())
        return std::nullopt;
    BlobRef r;
    r.id = j["blob"].get<std::string>();
    r.len = j["len"].get<std::uint64_t>();
    if (j.contains("type") && j["type"].is_string()) r.type = j["type"].get<std::string>();
    if (!uuid_bytes(r.id)) return std::nullopt;
    return r;
}

std::optional<std::array<std::uint8_t, 16>> uuid_bytes(std::string_view text) {
    if (text.size() != 36) return std::nullopt;
    std::array<std::uint8_t, 16> out{};
    std::size_t n = 0;
    for (std::size_t i = 0; i < text.size(); i++) {
        const char c = text[i];
        if (i == 8 || i == 13 || i == 18 || i == 23) {
            if (c != '-') return std::nullopt;
            continue;
        }
        const int hi = hex_value(c);
        const int lo = hex_value(text[++i]);
        if (hi < 0 || lo < 0 || n >= 16) return std::nullopt;
        out[n++] = static_cast<std::uint8_t>((hi << 4) | lo);
    }
    return n == 16 ? std::optional{out} : std::nullopt;
}

std::string uuid_text(const std::array<std::uint8_t, 16> &bytes) {
    static const char *hex = "0123456789abcdef";
    std::string s;
    s.reserve(36);
    for (std::size_t i = 0; i < 16; i++) {
        if (i == 4 || i == 6 || i == 8 || i == 10) s += '-';
        s += hex[bytes[i] >> 4];
        s += hex[bytes[i] & 0xf];
    }
    return s;
}

std::vector<std::byte> encode_chunk(std::string_view blob_id, std::uint64_t offset, std::uint64_t blob_len,
                                    std::span<const std::byte> payload) {
    const auto id = uuid_bytes(blob_id);
    if (!id) throw FjarrError("payload-invalid", "blob id is not a uuid: " + std::string(blob_id));
    if (payload.size() > MAX_CHUNK_PAYLOAD) throw FjarrError("payload-invalid", "blob chunk above 256 KiB (docs/08)");
    if (offset + payload.size() > blob_len) throw FjarrError("payload-invalid", "blob chunk overruns blob_len");
    std::vector<std::byte> frame(HEADER_BYTES + payload.size());
    frame[0] = static_cast<std::byte>(VERSION);
    for (std::size_t i = 0; i < 16; i++) frame[1 + i] = static_cast<std::byte>((*id)[i]);
    put_u64(frame.data() + 17, offset);
    put_u64(frame.data() + 25, blob_len);
    put_u32(frame.data() + 33, static_cast<std::uint32_t>(payload.size()));
    if (!payload.empty()) std::memcpy(frame.data() + HEADER_BYTES, payload.data(), payload.size());
    return frame;
}

std::optional<BlobChunk> parse_chunk(std::span<const std::byte> frame) {
    if (frame.size() < HEADER_BYTES) return std::nullopt;
    if (static_cast<std::uint8_t>(frame[0]) != VERSION) return std::nullopt;
    std::array<std::uint8_t, 16> id{};
    for (std::size_t i = 0; i < 16; i++) id[i] = static_cast<std::uint8_t>(frame[1 + i]);
    BlobChunk c;
    c.blob_id = uuid_text(id);
    c.offset = get_u64(frame.data() + 17);
    c.blob_len = get_u64(frame.data() + 25);
    const std::uint32_t payload_len = get_u32(frame.data() + 33);
    if (payload_len != frame.size() - HEADER_BYTES) return std::nullopt; // payload_len disagrees with the message
    if (payload_len > MAX_CHUNK_PAYLOAD) return std::nullopt;
    if (c.offset > c.blob_len || payload_len > c.blob_len - c.offset) return std::nullopt; // overruns blob_len
    c.payload = frame.subspan(HEADER_BYTES);
    return c;
}

// ------------------------------------------------------------ assembler

BlobAssembler::BlobAssembler(std::size_t max_blob, std::size_t pending_bytes, std::chrono::milliseconds ttl)
    : max_blob_(max_blob),
      limit_(pending_bytes),
      ttl_(ttl) {}

void BlobAssembler::erase(std::map<std::string, Entry>::iterator it) {
    bytes_ -= it->second.data.size();
    entries_.erase(it);
}

std::optional<std::string> BlobAssembler::on_chunk(const BlobChunk &chunk, clock::time_point now) {
    if (chunk.blob_len > max_blob_) {
        dropped_++;
        return std::nullopt;
    }
    auto it = entries_.find(chunk.blob_id);
    if (it == entries_.end()) {
        if (chunk.offset != 0) { // a blob that starts mid-way was already discarded
                                 // (or never seen): drop
            dropped_++;
            return std::nullopt;
        }
        it = entries_.emplace(chunk.blob_id, Entry{}).first;
        it->second.len = chunk.blob_len;
        it->second.first = now;
        it->second.data.reserve(static_cast<std::size_t>(chunk.blob_len));
    }
    Entry &e = it->second;
    if (e.complete || chunk.blob_len != e.len || chunk.offset != e.data.size()) { // a gap or a changed length: the blob is discarded
        dropped_++;
        erase(it);
        return std::nullopt;
    }
    e.data.append(reinterpret_cast<const char *>(chunk.payload.data()), chunk.payload.size());
    e.last = now;
    bytes_ += chunk.payload.size();
    e.complete = e.data.size() == e.len;
    // Bounded store: evict the oldest other blobs while over the limit (docs/08:
    // 8 MiB per channel).
    while (bytes_ > limit_ && entries_.size() > 1) {
        auto oldest = entries_.end();
        for (auto j = entries_.begin(); j != entries_.end(); ++j)
            if (j != it && (oldest == entries_.end() || j->second.first < oldest->second.first)) oldest = j;
        if (oldest == entries_.end()) break;
        dropped_++;
        erase(oldest);
    }
    return e.complete ? std::optional{chunk.blob_id} : std::nullopt;
}

bool BlobAssembler::complete(const std::string &id) const {
    auto it = entries_.find(id);
    return it != entries_.end() && it->second.complete;
}

std::optional<std::string> BlobAssembler::take(const std::string &id) {
    auto it = entries_.find(id);
    if (it == entries_.end() || !it->second.complete) return std::nullopt;
    std::string out = std::move(it->second.data);
    bytes_ -= out.size();
    entries_.erase(it);
    return out;
}

void BlobAssembler::discard(const std::string &id) {
    auto it = entries_.find(id);
    if (it != entries_.end()) erase(it);
}

void BlobAssembler::expire(clock::time_point now) {
    for (auto it = entries_.begin(); it != entries_.end();) {
        if (now - it->second.last > ttl_) {
            auto next = std::next(it);
            dropped_++;
            erase(it);
            it = next;
        } else {
            ++it;
        }
    }
}

} // namespace fjarr::blob
